packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -649,223 +600,326 @@ export function getValidatorApi(
-          strictFeeRecipientCheck: false,
-          // skip checking and recomputing head in these individual produce calls
-          skipHeadChecksAndUpdate: true,
-          commonBlockBody,
-          parentBlockRoot,
-        })
-      : Promise.reject(new Error("Builder disabled"));
-
-    const enginePromise = isEngineEnabled
-      ? produceEngineFullBlockOrContents(slot, randaoReveal, graffiti, {
-          feeRecipient,
-          strictFeeRecipientCheck,
-          // skip checking and recomputing head in these individual produce calls
-          skipHeadChecksAndUpdate: true,
-          commonBlockBody,
-          parentBlockRoot,
-        }).then((engineBlock) => {
-          // Once the engine returns a block, in the event of either:
-          // - suspected builder censorship
-          // - builder boost factor set to 0 or builder selection `executionalways`
-          // we don't need to wait for builder block as engine block will always be selected
-          if (
-            engineBlock.shouldOverrideBuilder ||
-            builderBoostFactor === BigInt(0) ||
-            builderSelection === routes.validator.BuilderSelection.ExecutionAlways
-          ) {
-            controller.abort();
-          }
-          return engineBlock;
-        })
-      : Promise.reject(new Error("Engine disabled"));
-
-    const [builder, engine] = await resolveOrRacePromises([builderPromise, enginePromise], {
-      resolveTimeoutMs: cutoffMs,
-      raceTimeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
-      signal: controller.signal,
-    });
-
-    if (builder.status === "pending" && engine.status === "pending") {
-      throw Error("Builder and engine both failed to produce the block within timeout");
-    }
-
-    if (isEngineEnabled) {
-      if (engine.status === "rejected") {
-        logger.warn(
-          "Engine failed to produce the block",
-          {
-            ...loggerContext,
-            durationMs: engine.durationMs,
-          },
-          engine.reason
-        );
-      } else if (engine.status === "pending") {
-        logger.warn("Engine failed to produce the block within cutoff time", {
-          ...loggerContext,
-          cutoffMs,
-        });
-      }
-    }
-
-    if (isBuilderEnabled) {
-      if (builder.status === "rejected") {
-        if (builder.reason instanceof NoBidReceived) {
-          logger.info("Builder did not provide a bid", {
-            ...loggerContext,
-            durationMs: builder.durationMs,
-          });
-        } else {
-          logger.warn(
-            "Builder failed to produce the block",
-            {
-              ...loggerContext,
-              durationMs: builder.durationMs,
-            },
-            builder.reason
-          );
+    const [commonBlockBody, [builder, engine]] = await Promise.all([
+      produceCommonBlockBody(blockAttributes, currentState, loggerContext),
```

Comment:
`produceCommonBlockBody` needs to stay last due to some synced code inside
same for the other 2 calls